### ISSUE TITLE
High CPU Usage in WmiPrvSE.exe observed with PIA Split Tunnel

### ISSUE THREAD

**Issue by std-move** [pia-foss/desktop#62]
Hello,

I have noticed that `WmiPrvSE.exe`'s CPU usage is relatively high when `Private Internet Access Service` is running and Split Tunnel is enabled (there is at least one split tunnel rule configured). PIA looks to make heavy use of WMI.

The usage seems to vary, but is generally around 1/5 of a performance core on an Intel 13th gen laptop CPU. That might not seem like much, but makes a big difference to thermals, fan activity and power usage.

To put things in perspective, when monitoring usages via `resmon` over time, `WmiPrvSE.exe` dominates. "Idle" browsers like Firefox and Chrome, with plenty of tabs open, seem to use less CPU in total (yes, I am aware of multiprocessing) even though there's plenty of background activity.

Observed on Win10 22H2/PIA 3.3.1 and Win11 22H2/PIA 3.5.1.

Is there a way something could done about this?

**Comment by std-move** [comment]
Hello, I don't see any significant difference between 'Follow app rules' and 'VPN DNS only', CPU usage seems to be high in either case.

**Comment by kp-john-mair** [comment]
Thanks guys - sorry about that. Unfortunately we don't have any time allotted to look into the windows split tunnel for a while, until about Q2 or Q3 of 2024 at the earliest. But when i get time, i'll jump in. Feel free to add more info to this ticket as you discover it.

**Comment by marcelxl** [comment]
Same problem here.
PIA x64 - 3.5.3-07926
Windows 11 23H2 , Intel core i5

With the WMI Provider host using about 6% to 10% CPU usage while idle, the performance loss (and sluggish of the system) is significant.

Suggested changes did not change the behavior, only stopping the PIA NT service does.

**Comment by ossnorry** [comment]
Issue is still sucking up 40% of my CPU, any updates on this?

**Comment by Code-Vlue** [comment]
Issue is still present. WMI service has variable CPU usage while split tunneling.

**Comment by Code-Vlue** [comment]
Issue is still present. When Split Tunnel is enabled and Administrator Rights prompt is accepted immediately get WMI service CPU usage which generally idles at around 6%. Usage will persist through restart even if Split Tunneling is tuned off in settings. Once the original admin prompt has been accepted the only way to remove WMI usage is to uninstall PIA all together and reinstall making sure you don't enable split tunneling again or you will have to reinstall again.

**Comment by Code-Vlue** [comment]
Update: Found a patchy work around, if you run the winmgmt /resetrepository command in the CMD as admin then the problem goes away however it seems like you need to do it on every boot. I made a boot task to run it every time but seems like the issue might be due to PIA causing WMI corruption due to split tunneling.

**Comment by Average-GH-Enjoyer** [comment]
![image](https://github.com/user-attachments/assets/b1467584-52cf-4ed6-bad5-ea014cbfffc9)

I can confirm it is present on my PC as well. I couldn't figure out which app was causing this and had to reinstall Windows to find out. It's very frustrating. With a 7% CPU usage on my laptop, it doesn't do any good for the battery life.

**Comment by sl4m4j4m4** [comment]
just signed up for PIA and finding out this is an issue that has apparently been going on since 2023. considering getting a refund. this is a huge waste of laptop battery.

**Comment by mibby** [comment]
@kp-john-mair Also came across this problem and finally diagnosed it down to PIA erroneously calling WMI. Hoping we see this worked on eventually. Uninstalling PIA client entirely resolves the issue, as well as stopping all PIA services. A functional workaround in the meantime is restarting the Windows Management Instrumentation service after startup, but this is not ideal.

PIA Client Windows x64 3.6.1-08339 + Split Tunnel
Windows 11 (OS Build 27842.1000)

**Comment by ossnorry** [comment]
This is still an issue, does PIA intended to fix it eventually? I hate that we're being ignored on this major issue, two years in. 

@marcelxl Has anyone attempted to build the client without that code issue? I hate asking questions like that, it's just crazy how long this issue has persisted.

**Comment by ossnorry** [comment]
Still waiting for an update from PIA on this, it feels like they're ignoring us when this is a bit of a huge problem...

**Comment by kp-john-mair** [comment]
Hi guys, sorry about lack of response on this issue. We intend to look into and fix this when we get a chance - but unfortunately priorities are elsewhere for now. Hopefully we'll have resources in the near-ish future to get back to this and we'll let you know when we're on it! Thanks for looking into this, all very useful information!

**Comment by mibby** [comment]
How soon is "near-ish"? The desktop app hasn't received an update in 10 months and this issue was reported nearly 2 years ago. I'd very much like to use the split tunnel again but feeling like a neglected customer of PIA. Many dependencies are also outdated. (OpenVPN, WireGuard, Unbound, etc.)

The app monitoring could be rewritten to act as an ETW consumer using push events instead of WMI polling if you did not want to create a kernel driver for this.

**Comment by mibby** [comment]
Glad to finally see movement with a new desktop version released, but unfortunately didn't update any of the main dependencies or address this issue. Perhaps next year I suppose!

**Comment by Code-Vlue** [comment]
A bit disappointed to come back after all this time to PIA only to see this is still an issue. Seems to be a trend that VPN services keep disappointing me lately.

**Comment by ciarpame** [comment]
A year ago, I canceled my subscription because this annoying bug (which drove me crazy and wasted a lot of my time before I understood the cause) was not fixed. So, for a year, I subscribed to another VPN. 
I have now checked again, but nothing has changed. I understand that PIA has other priorities, which unfortunately do not match mine. It's a shame because I prefer PIA for a lot of reasons (including the fact that it doesn't constantly insist on me buying additional services like the other VPN service does), but this issue is unfortunately a deal breaker for me.
I really hope to read some good news when I will check back next year.

**Comment by ossnorry** [comment]
This issue is still present, won't someone from PIA say something? We're 2+ years into a performance breaking issue on many systems, and we've still had no fix. 
@MysticRyuujin @JonathonH-PIA @jacobly0

**Comment by ossnorry** [comment]
Is anyone part of this thread able to commit a fix by chance?

**Comment by mibby** [comment]
@kp-john-mair Where are those promised resources from the "near-ish future"? It's 2026 and this issue is still not addressed. Took 5 months just to _publicly release_ the 3.7.0 commit with minimal changes and yet still, no one bothered with any dependency updates.

PIA Windows client moves at a non-existent pace while other VPN companies are pushing changes early and often. Think it's time I drop this nonsense and find an alternative that actually values their customers. I will continue to inform public discussions around VPNs / PIA of this severe drawback until this issue is actually fixed.

**Comment by Johnnycyan** [comment]
This is absolutely stupid. It took me FAR too long to realize this was the process responsible for this mess. I've been going crazy trying to figure this out. You should feel ashamed to have not only allowed this to exist in the first place but continue allowing it to exist for almost 3 years after being reported to you.

Any developer worth their salt could have told you that polling for every single process on the computer every 100ms is absolutely stupid.

**Comment by aaronth07** [comment]
Absolute insanity. I will not be renewing my subscription another term, which expires in a few months. This plus the ton of other "minor" issues that end up stacking up over time (e.g. being completely unusable on Mac disabling system services and reinstalling every time you want to toggle on/off), plus most IPs being extremely slow (going from 1gbps to 20mbps down) and even more restricted than ever, the value is just not there. 

Of course, I still have to stick with it until my subscription is up. Does anybody know if there is perhaps a way we can limit the polling rate of an application to a Windows function/service such as this?

**Comment by ossnorry** [comment]
Still waiting on a response for this, been many years now and effects literally tens of thousands of people and is nuking their PC performance and battery life. 


Please just acknowledge the issue, it's an easy fix.

**Comment by ossnorry** [comment]
Have anyone here considered alternate VPN clients with PIA, are the are that support PIA with port forwarding and split tunneling?

**Comment by ossnorry** [comment]
> Still waiting on a response for this, been many years now and effects literally tens of thousands of people and is nuking their PC performance and battery life.
> 
> Please just acknowledge the issue, it's an easy fix.

@MysticRyuujin @JonathonH-PIA @jacobly0 
Any word on a fix for this long-standing issue ruining the PCs of tens of thousands of users?

**Comment by MysticRyuujin** [comment]
I don't work for PIA, as far as I can tell they've completely abandoned their open source repos 🤷🏻‍♂️

**Comment by ossnorry** [comment]
> I don't work for PIA, as far as I can tell they've completely abandoned their open source repos 🤷🏻‍♂️

The problem is they still ship the software, and tell people how great it is. But have neglected to actually update the software too, so it's way out of date in terms of security, features, and the bugs outlined here an in issues.

**Comment by ossnorry** [comment]
Does anyone know if there's an alternate client that accepts PIA creds that isn't the worst thing ever to set up, and still supports port forwarding and split-tunnel?

**Comment by aaronth07** [comment]
> Does anyone know if there's an alternate client that accepts PIA creds that isn't the worst thing ever to set up, and still supports port forwarding and split-tunnel?

Nope, and my renew is coming up in one month. Going to cancel my subscription rather than renewing.

**Comment by ossnorry** [comment]
I'm not even sure if anyone should waste time trying to fix it. We don't even know if PIA would take the pull request, they've been basically radio silent on all the issues for 3 dang years. it's insane. Is anyone even working there?
